client/components/popover/index.jsx

```diff
@@ -81,245 +86,176 @@ class Popover extends Component {
 	}
 
 	// --- ESC key ---
 	bindEscKeyListener() {
-		if ( ! this.props.closeOnEsc ) {
-			return null;
+		if ( this.props.closeOnEsc ) {
+			document.addEventListener( 'keydown', this.onKeydown, true );
 		}
-
-		this.debug( 'adding escKey listener ...' );
-		document.addEventListener( 'keydown', this.onKeydown, true );
 	}
 
 	unbindEscKeyListener() {
-		if ( ! this.props.closeOnEsc ) {
-			return null;
+		if ( this.props.closeOnEsc ) {
+			document.removeEventListener( 'keydown', this.onKeydown, true );
 		}
-
-		this.debug( 'unbinding `escKey` listener ...' );
-		document.removeEventListener( 'keydown', this.onKeydown, true );
 	}
 
-	onKeydown( event ) {
-		if ( event.keyCode !== 27 ) {
-			return null;
-		}
+	onKeydown = event => {
+		if ( event.keyCode === 27 ) {
+			const domContext = ReactDom.findDOMNode( this.props.context );
+			if ( domContext ) {
+				debug( 'Refocusing the previous active DOM node' );
+				domContext.focus();
+			}
 
-		if ( this.domContext ) {
-			this.debug( 'Refocusing the previous active DOM node' );
-			this.domContext.focus();
+			this.close( true );
 		}
-
-		this.close( true );
-	}
+	};
 
 	// --- click outside ---
-	bindClickoutHandler( el = this.domContainer ) {
-		if ( ! el ) {
-			this.debug( 'no element to bind clickout ' );
-			return null;
-		}
-
-		if ( this._clickoutHandlerReference ) {
-			this.debug( 'clickout event already bound' );
-			return null;
-		}
-
-		this.debug( 'binding `clickout` event' );
-		this._clickoutHandlerReference = clickOutside( el, this.onClickout );
+	bindClickoutHandler() {
+		document.addEventListener( 'click', this.onClickout );
 	}
 
 	unbindClickoutHandler() {
-		if ( this._clickoutHandlerReference ) {
-			this.debug( 'unbinding `clickout` listener ...' );
-			this._clickoutHandlerReference();
-			this._clickoutHandlerReference = null;
-		}
+		document.removeEventListener( 'click', this.onClickout );
 	}
 
-	onClickout( event ) {
-		let shouldClose =
-			this.domContext && this.domContext.contains && ! this.domContext.contains( event.target );
+	onClickout = event => {
+		const popoverContext = this.popoverInnerNodeRef.current;
+		let shouldClose = popoverContext && ! popoverContext.contains( event.target );
+
+		if ( shouldClose && this.props.context ) {
+			const domContext = ReactDom.findDOMNode( this.props.context );
+			shouldClose = domContext && ! domContext.contains( event.target );
+		}
 
-		if ( this.props.ignoreContext && shouldClose ) {
+		if ( shouldClose && this.props.ignoreContext ) {
 			const ignoreContext = ReactDom.findDOMNode( this.props.ignoreContext );
-			shouldClose =
-				shouldClose &&
-				ignoreContext &&
-				ignoreContext.contains &&
-				! ignoreContext.contains( event.target );
+			shouldClose = ignoreContext && ! ignoreContext.contains( event.target );
 		}
 
 		if ( shouldClose ) {
 			this.close();
 		}
-	}
+	};
 
 	// --- window `scroll` and `resize` ---
-	bindDebouncedReposition() {
+	bindReposition() {
```

Comment:
It like needs to be repositioned to ensure the popover remains on screen, if possible. As an element moves towards the top of the screen, the popover may need to be moved below the element to remain in view.

Reply:
The popover element is a child of `body`, and is absolutely positioned relative to the viewport. Not relative to the element that is scrolling. That's why it needs to be repositioned on every scroll. I agree that's suboptimal and that we could position the popover in such a way that it scrolls together with the content that it's attached to. But that's for another awesome PR 😉

Reply:
Sounds like an excellent enhancement, if it's doable 🙂

Reply:
Now, when testing, I discovered that it's the other way around: the whole `body` is scrolling, so popovers are always properly attached even without the reposition scroll handler.

It's the static elements, sidebar and masterbar, that need this onscroll repositioning.

Reply:
Ah, that's interesting! This definitely complicates things.

If I were designing this component from scratch, I'd probably have the component require a prop that indicates whether to attach itself to the body or to the window. This would be used to toggle some CSS classes that either made it `position: absolute` or `position: fixed`, with the portal as a direct child of the body. That way neither scenario would require scroll listeners, but as a downside it would require the consumer to know what it wants.

Still, that's a discussion for a future PR, if ever 🙂